### PR TITLE
Trilinos: update SuperLU dependency

### PR DESCRIPTION
**Pull request by lucbv**
SuperLu has deprecated versions below 5.2.1 which includes 4.3 currently used as default in Trilinos. This will trigger spec failures in the next spack release and already creates issues with spack/develop. The fix bumps the minimum version of superlu and creates conflicts with amesos and ifpack as they do not support the 5.2.1 interface.

### CODE REVIEW

File: var/spack/repos/builtin/packages/trilinos/package.py
Comment (psakievich) [inline var/spack/repos/builtin/packages/trilinos/package.py]:
```suggestion
    conflicts("+amesos", when="+superlu@:4.3")
    conflicts("+ifpack", when="+superlu@:4.3")
```
Should we consider embedding this knowledge by making the conflict conditional on the version rather than just turning it off? Would this be the right version?

File: var/spack/repos/builtin/packages/trilinos/package.py
Comment (psakievich) [inline var/spack/repos/builtin/packages/trilinos/package.py]:
```suggestion
    depends_on("superlu@4.3: +pic", when="+superlu")
```
Also would this be valid?  If trilinos can support more than one version it would be best to embed that knowledge in the package

File: var/spack/repos/builtin/packages/trilinos/package.py
Comment (lucbv) [reply var/spack/repos/builtin/packages/trilinos/package.py]: We would need `when="+superlu@5.2.1:"` to turn off the packages when using versions of superlu at 5.2.1 and above.

File: var/spack/repos/builtin/packages/trilinos/package.py
Comment (lucbv) [reply var/spack/repos/builtin/packages/trilinos/package.py]: The problem is this is not "Spack" valid in the sense that you cannot concretize with version 4.3 in the currently develop version of Spack. Actually this probably means that we want to get rid of the code path that supports 4.3 in Trilinos directly but that's a different level of effort.

Comment (psakievich) [reply var/spack/repos/builtin/packages/trilinos/package.py]: You can if you set `config:deprecated:true`, or if the users define a `superlu` package that doesn't have 4.3 deprecated.

I would prefer our package recipe be written based on what the code **can** successfully compile against, and allow the spack solver to pick non-deprecated versions. 

An exception to this would be security concerns.

If we had written our `superlu` dependency as a range from the start then the concretization failure wouldn't have occurred when the `superlu` version was deprecated.

**Comment by sebrowne** [comment]
My two cents: Agree with Phil.  If we want to be a bit more picky about it and encourage people to use the newer SuperLU for newer Trilinos, we could change the depends on as follows:

```
    depends_on("superlu@4.3: +pic", when="@:15.1.0+superlu")
    depends_on("superlu@5.2.2: +pic", when="@15.1.1:+superlu")
```

I think that will still fix the concretization, no?  When you combine it with the conflict() for amesos and ifpack?  It does prevent building amesos and ifpack with newer Trilinos releases (not sure that's a bad thing given that they are deprecated packages and will be removed next year).

**Comment by sethrj** [comment]
> I would prefer our package recipe be written based on what the code can successfully compile against, and allow the spack solver to pick non-deprecated versions.

100% agree. If the latest trilinos builds with 4.3 (even though it's deprecated) we shouldn't prohibit that. We just need to extend the version range so that it's forward compatible, and add conflicts for the packages/version ranges that don't support the newer SuperLU.

File: var/spack/repos/builtin/packages/trilinos/package.py
Comment (sethrj) [reply var/spack/repos/builtin/packages/trilinos/package.py]:
```suggestion
    conflicts("+amesos", when="+superlu@5:")
    conflicts("+ifpack", when="+superlu@5:")
```

It should probably be something like this. If an older superlu is enabled then these old packages should still build with it. These are going to go away eventually when epetra is removed, right?

File: var/spack/repos/builtin/packages/trilinos/package.py
Comment (sethrj) [reply var/spack/repos/builtin/packages/trilinos/package.py]:
```suggestion
    depends_on("superlu@4.3: +pic", when="+superlu")
```
Should definitely be this.

**Review by sethrj — CHANGES_REQUESTED** [review]

**Comment by lucbv** [comment]
So my thinking is that the `package.py` in a given version of Spack should not by default depend on deprecated package.
If users want to do that they should get an older version of Spack when the package was not deprecated or they should request the deprecated package specifically from the command line with `--deprecated +superlu@4.3`

### Note:
once again, we are **not** talking about what Trilinos supports but rather what Spack supports. Once the current develop of Spack gets tagged into a new version of Spack there will be failures if we have `depends_on("superlu@4.3: +pic", when="@:15.1.0+superlu")` because **Spack** does not support it by default

**Comment by sethrj** [comment]
> So my thinking is that the `package.py` in a given version of Spack should not by default depend on deprecated package. If users want to do that they should get an older version of Spack when the package was not deprecated or they should request the deprecated package specifically from the command line with `--deprecated +superlu@4.3`
> 
> once again, we are **not** talking about what Trilinos supports but rather what Spack supports. Once the current develop of Spack gets tagged into a new version of Spack there will be failures if we have `depends_on("superlu@4.3: +pic", when="@:15.1.0+superlu")` because **Spack** does not support it by default

@lucbv `depends_on("superlu@4.3:")` should allow newer versions to work; it should by default avoid the older versions. Did you get a failure while testing that?

**Comment by lucbv** [comment]
Okay after further thinking here are a few thoughts

1. I am not sure deprecating a version of a package makes sense, i.e. why is superlu allowed to say that it's source code at 4.3 no longer works? Surely if it has all the features a downstream application requires they should be allowed to use it unless there is a security concern?
2. We all agree that at least the latest Trilinos in develop should be made to build so we want that one to target some 5.2 version of SuperLU
3. That leaves the questions of older version of Trilinos. Here is the changed I made locally:
```
    depends_on("superlu@5.2.2 +pic", when="@15.0.0:+superlu")
    depends_on("superlu@4.3 +pic", when="@:14.4.0+superlu")
```
After which I get the following behavior:

```
$ spack spec trilinos@14.4.0 %gcc@11.4.1 +mpi+exodus~epetra+tpetra+muelu+belos+ifpack2+amesos2+zoltan+stk+boost+superlu+hdf5+shards
==> Error: concretization failed for the following reasons:

   1. Cannot satisfy 'superlu@4.3'
   2. Cannot satisfy 'superlu@4.3'
        required because trilinos depends on superlu@4.3+pic when @:14.4.0+superlu 
          required because trilinos@14.4.0%gcc@11.4.1+amesos2+belos+boost~epetra+exodus+hdf5+ifpack2+mpi+muelu+shards+stk+superlu+tpetra+zoltan requested explicitly 
   3. Cannot satisfy 'superlu@5:' and 'superlu@4.3
        required because trilinos depends on superlu@4.3+pic when @:14.4.0+superlu 
          required because trilinos@14.4.0%gcc@11.4.1+amesos2+belos+boost~epetra+exodus+hdf5+ifpack2+mpi+muelu+shards+stk+superlu+tpetra+zoltan requested explicitly 
        required because superlu variant build_system value cmake when [, , ] 
          required because trilinos depends on superlu@4.3+pic when @:14.4.0+superlu 
            required because trilinos@14.4.0%gcc@11.4.1+amesos2+belos+boost~epetra+exodus+hdf5+ifpack2+mpi+muelu+shards+stk+superlu+tpetra+zoltan requested explicitly 
```
That is what I expect and the way to work around this is to add `--deprecated` to the spack command line:
```
$ spack spec --deprecated trilinos@14.4.0 %gcc@11.4.1 +mpi+exodus~epetra+tpetra+muelu+belos+ifpack2+amesos2+zoltan+stk+boost+superlu+hdf5+shards
==> Warning: using "superlu@4.3" which is a deprecated version
Input spec
--------------------------------
 -   trilinos@14.4.0%gcc@11.4.1+amesos2+belos+boost~epetra+exodus+hdf5+ifpack2+mpi+muelu+shards+stk+superlu+tpetra+zoltan

Concretized
--------------------------------
 -   trilinos@14.4.0%gcc@11.4.1~adelus~adios2~amesos+amesos2+anasazi~aztec~basker+belos+boost~chaco~complex~cuda~cuda_rdc~debug~dtk~epetra~epetraext~epetraextbtf~epetraextexperimental~epetraextgraphreorderings+exodus+explicit_template_instantiation~float+fortran~gtest+hdf5~hypre~ifpack+ifpack2~intrepid~intrepid2~ipo~isorropia+kokkos~mesquite~minitensor~ml+mpi+muelu~mumps~nox~openmp~pamgen~panzer~phalanx~piro~python~rocm~rocm_rdc~rol~rythmos+sacado~scorec+shards+shared~shylu+stk~stokhos~stratimikos~strumpack~suite-sparse+superlu~superlu-dist~teko~tempus~test~thyra+tpetra~trilinoscouplings~wrapper~x11+zoltan~zoltan2 build_system=cmake build_type=Release cxxstd=17 generator=make gotype=long_long arch=linux-rhel9-broadwell
```

**Comment by kuberry** [comment]
> Okay after further thinking here are a few thoughts
> 
>     1. I am not sure deprecating a version of a package makes sense, i.e. why is superlu allowed to say that it's source code at 4.3 no longer works? Surely if it has all the features a downstream application requires they should be allowed to use it unless there is a security concern?
> 

I agree with you @lucbv about deprecation being used sparingly (security concerns). Otherwise, every package can just mark all older versions 'deprecated' and cause concretization errors upstream, and the use of the `--deprecated` flag becomes needed in a huge number of scenarios.  The worst part is that the failure to concretize error doesn't even point at the deprecation in superlu, but rather that `superlu@4.3` can't be satisfied. Many people using spack would be unlikely to realize that they only need to add the `--deprecated` flag and would assume they simply couldn't install Trilinos at that version.

**Comment by sethrj** [comment]
@lucbv Can you describe the matrix of Trilinos versions/packages vs SuperLU versions (regardless of the "deprecated" flag)? I really think the answer is going to be something like:
```c++
    depends_on("superlu@4.3: +pic", when="+superlu")
    conflicts("+amesos", when="+superlu@5:")
    conflicts("+ifpack", when="+superlu@5:")
```

**Comment by kuberry** [comment]
> @lucbv Can you describe the matrix of Trilinos versions/packages vs SuperLU versions (regardless of the "deprecated" flag)? I really think the answer is going to be something like:
> 
> ```c++
>     depends_on("superlu@4.3: +pic", when="+superlu")
>     conflicts("+amesos", when="+superlu@5:")
>     conflicts("+ifpack", when="+superlu@5:")
> ```

This will have the effect of:
`spack spec trilinos@14.4.0 +mpi+exodus~epetra+tpetra+muelu+belos+ifpack2+amesos2+zoltan+stk+boost+superlu+hdf5+shards`
concretizing with `superlu@5.3` and disabling `ifpack` and `amesos` for `trilinos14.4.0`. Also, if `+ifpack+amesos+epetra` were specified, this would also result in a concretization error (because the only version available to concretize super-lu to is 5.3).

**Comment by psakievich** [comment]
@lucbv @kuberry deprecation is also a way for the superlu maintainers to say they have no intention of continuing to maintain that version. See [docs](https://spack.readthedocs.io/en/latest/packaging_guide.html#deprecating-old-versions).  I think this is the intended scenario even though it is uncomfortable.  Essentially this communicates to the trilinos team they are relying on unmaintained software and they should either 

a) update their own software as a result
b) offer to maintain the version they rely on 
c) find another way to vendor it (really we never want to do this)


I was going to ping the superlu maintainers in this comment, but interestingly there isn't one: https://packages.spack.io/package.html?name=superlu

Any thoughts on removing the deprecation @sethrj?

**Comment by psakievich** [comment]
>  If an older superlu is enabled then these old packages should still build with it. These are going to go away eventually when epetra is removed, right?

Although for this specific instance maybe we want the `--deprecated` flag in place if these are `eptra` dependent packages and they are going to be removed from `trilinos` soon anyways.

**Comment by lucbv** [comment]
@psakievich 
Not supporting old release is pretty common actually the whole idea behind freezing a set of source code and making it a release is more or less saying that it won't be changed moving forward and that newer releases will now be where new features and fixes will go. I don't think that deprecating an old release is really appropriate? Is SuperLU going to delete the tag from their repo and delete the associated release download link?

**Comment by lucbv** [comment]
@sethrj 
```
    depends_on("superlu@4.3: +pic", when="+superlu")
    conflicts("+amesos", when="+superlu@5:")
    conflicts("+ifpack", when="+superlu@5:")
```
Reasonably captures the fact that `amesos` and `ifpack` should be turned off when a superlu version above 4.3 is used so I don't mind adding that but it does not solve the initial problem I raised which is: Trilinos currently does not concretize properly so we should decide on what the new default superlu should be? We can all agree that having a package that fails to build by default is not appropriate!

**Comment by psakievich** [comment]
> @psakievich Not supporting old release is pretty common actually the whole idea behind freezing a set of source code and making it a release is more or less saying that it won't be changed moving forward and that newer releases will now be where new features and fixes will go. I don't think that deprecating an old release is really appropriate? Is SuperLU going to delete the tag from their repo and delete the associated release download link?

@lucbv yeah well my response was mostly a thought exercise since `superlu` does not have an actual maintainer.  I am not sure deprecation is appropriate in this instance either. I think that could be remove as part of this PR as well, but I also think the modifications here are appropriate. We should be handling our `superlu` version constraints with ranges.

@sethrj you made the deprecation mark in spack, would you be opposed to undeprecating it?

**Comment by sethrj** [comment]
Here are my suggestions:
1. Un-deprecate SuperLU 4.3. That was my mistake and if it's still needed to use some packages, and if it's still available to be installed, then we should be able to build those packages against it.
2. Widen the allowed version ranges. I don't think `4.3` is exactly the right version range; it should probably be `@4`. Likewise with the newer version: there's no way that `5.2.2` works and `5.2.1` or `5.2.3` don't.
3. Mark conflicts that we know about. If we know `+amesos+superlu` fails for `@5.2` and passes for `@4.3`, then add a conflict `"superlu@:4", when="+amesos+superlu"` and add a comment for the exact versions that pass and fail.

**Comment by lucbv** [comment]
Looking at the change proposed by @sethrj more carefully, since it specifies a range it should work fine as long as the solver picks a version supported by Trilinos.
Looking at this file in the trilinos repo: [cmake/TPLs/FindTPLSuperLU.cmake](https://github.com/trilinos/Trilinos/blob/master/cmake/TPLs/FindTPLSuperLU.cmake)
I can draw up the following table that shows when a new version of SuperLu was added to Trilinos:

| Trilinos | SuperLU | package.py |
| --- | --- | --- |
| 12.18.1 | 5.2.1 | depends_on("superlu@3:5.2" when="@12.18.1: +superlu") |
| 12.14.1 | 5.0 and 5.1.1 | depends_on("superlu@3:5.1.1" when="@12.14.1 +superlu") |
| 10.8.0 | 4.1 | depends_on("superlu@3:4" when="@:12.12.1 +superlu") |
| 10.4.1 | 4.0 | N/A |
| 10.0.0 | 3.0 | N/A |

Currently we do not have more explicit support beyond 5.2.1 in that particular TPL file, that does not mean there is an issue building against newer SuperLU, I just don't have easy information to assess it...

**Comment by lucbv** [comment]
Please feel free to amend the table above to reflect the syntax you want for the dependency statement in the third column.
Then for the amesos and ifpack conflicts I like the following:
```
conflict("+amesos", when="+superlu@5:")
conflict("+ifpack", when="+superlu@5:")
```
as it captures well that these two packages cannot build against versions of SuperLU higher than 4.3.
If we could settle this today that would be nice so we can have CI run by tomorrow morning leading to a proper resolution this week : )

**Comment by sebrowne** [comment]
I think the last approach proposed by @lucbv is the one I would suggest.  Do we still have to un-deprecate 4.3, or will it concretize the default version with a supported SuperLU?  That's what I hope happens with the concretizer, but I haven't checked it.  I don't see a huge issue with having to pass `--deprecated` for the older versions of Trilinos.

**Comment by psakievich** [comment]
Seems reasonable to me.

**Comment by psakievich** [comment]
@lucbv will you remove the deprecation for `superlu`? You can do it quickly with `spack edit superlu` and just remove the deprecation tag for that version.

**Comment by lucbv** [comment]
@sethrj @psakievich 
This should now include all the changes we want including removing deprecation in superlu.
Let me know if this looks fine, it looks like the pipeline is almost finished and passing...

**Review by sethrj — APPROVED** [review]
LGTM! Thanks.

**Review by kuberry — APPROVED** [review]

**Comment by kuberry** [comment]
LGTM too. Thanks @lucbv  

This is an aside, but one thing I see as an issue going forward is how to pin dependencies w.r.t. CI testing. Now we will pin to superlu up to a max of 5.2, but we could have instead done `depends_on("superlu@3:", when="@12.18.1: +superlu")`. We pin Kokkos to an exact version (for a good reason, but could someday become a range), we pin this to a range that we've tested with CI, and for many other dependencies we provide an open ended range (which are not under CI testing).

**Review by psakievich — APPROVED** [review]

**Pull request merged**
